### PR TITLE
Sprint4/feature/us 3.2 location class

### LINKED ISSUE #13: US-3.2: Classroom location from schedule data
As a user, I want the app to determine my classroom locations from course data for accurate directions.

**Acceptance criteria:**

- [ ] The app can get classroom locations (campus, building, floor, classroom number) from course data.

### LINKED ISSUE #229: AT-3.2: Classroom location from schedule data
**Acceptance criteria:**

- [ ] The app can get classroom locations (campus, building, floor, classroom number) from course data.

### PR DESCRIPTION
**Pull request by philniko**
- Persist course schedule for future "directions to next class" feature

 - Stores parsed Google Calendar course data (campus, building, room) to AsyncStorage so it can be retrieved across sessions. Adds getNextClass() to look up the user's next upcoming class — groundwork for automatically routing the user to their next course location.

 - Also fixes broken test suites (parseCourseEvents, Schedule) by adding a global AsyncStorage mock, and adds missing test coverage for saveSchedule, loadCachedSchedule, getNextClass, pagination in the Calendar API, and token validation edge cases.

Closes #13 
Closes #229

### CODE REVIEW

File: __tests__/GoogleCalendarService.test.tsx [hunk 1]
```diff
@@ -118,6 +118,20 @@ describe("services/GoogleCalendarService", () => {
         "Calendar list failed (500): Server Error",
       );
     });
+
+    it("throws for a non-string token", async () => {
```
Comment (khalillabban) [inline __tests__/GoogleCalendarService.test.tsx]: These token-validation test cases appear again later for `fetchCalendarEventsInRange`. It might be worth extracting a small shared helper or grouping them in a shared describe block so the validation expectations stay consistent if the error messages change later.

File: __tests__/GoogleCalendarService.test.tsx [hunk 2]
```diff
@@ -292,5 +306,59 @@ describe("services/GoogleCalendarService", () => {
...
+        timeMax: new Date("2026-06-01T00:00:00.000Z"),
+      });
+
+      expect(items).toHaveLength(3);
+      expect(items.map((i) => i.id)).toEqual(["e1", "e2", "e3"]);
+      expect(global.fetch).toHaveBeenCalledTimes(2);
+
+      // Second call should include the pageToken
+      const [secondUrl] = (global.fetch as jest.Mock).mock.calls[1];
+      expect(secondUrl).toContain("pageToken=token-page-2");
```
Comment (khalillabban) [inline __tests__/GoogleCalendarService.test.tsx]: Minor suggestion: instead of checking a raw substring here, you could parse the URL and assert on the query parameter (e.g., using URLSearchParams). That would make the test a bit more robust against formatting changes in the request URL.

File: __tests__/parseCourseEvents.test.tsx [hunk 3]
```diff
@@ -19,26 +34,62 @@ describe("parseCourseEvents", () => {
...
+      },
+    ];
+
+    const res = parseCourseEvents(events);
+    expect(res[0].campus).toBe("");
+    expect(res[0].building).toBe("");
+    expect(res[0].room).toBe("");
+  });
+
+  it("supports all-day events using date", () => {
```
Comment (khalillabban) [inline __tests__/parseCourseEvents.test.tsx]: Suggestion: it may be worth clarifying whether all-day Google Calendar events should really be treated as classes here. For the “next class / directions” feature, including all-day events in the parsed schedule could cause unexpected routing behavior later unless they’re filtered upstream.

File: __tests__/parseCourseEvents.test.tsx [hunk 4]
```diff
@@ -52,11 +103,10 @@ describe("parseCourseEvents", () => {
       },
     ];
 
-    const res = parseCourseEvents(events);
-    expect(res).toHaveLength(0);
+    expect(parseCourseEvents(events)).toHaveLength(0);
   });
 
-  it("filters events with invalid dates", () => {
+  it("filters events with invalid start date", () => {
```
Comment (khalillabban) [inline __tests__/parseCourseEvents.test.tsx]: We cover an invalid start date here, but I don’t see the symmetric case for an invalid end date. Adding that test would help lock down the parser’s behavior a bit more completely.

File: __tests__/Schedule.test.tsx [hunk 5]
```diff
@@ -26,8 +26,14 @@ jest.mock("../services/GoogleCalendarService", () => ({
 }));
 
 const mockParseCourseEvents = jest.fn();
+const mockLoadCachedSchedule = jest.fn();
+const mockSaveSchedule = jest.fn();
+const mockGetNextClass = jest.fn();
 jest.mock("../utils/parseCourseEvents", () => ({
   parseCourseEvents: (...args: any[]) => mockParseCourseEvents(...args),
```
Comment (khalillabban) [inline __tests__/Schedule.test.tsx]: Small nit: using any[] here bypasses the type guarantees from parseCourseEvents. If possible, using the real function signature (or Parameters<typeof parseCourseEvents>) would keep the mock aligned with the implementation and prevent accidental argument mismatches later.

File: __tests__/Schedule.test.tsx [hunk 6]
```diff
@@ -26,8 +26,14 @@ jest.mock("../services/GoogleCalendarService", () => ({
...
+  loadCachedSchedule: (...args: any[]) => mockLoadCachedSchedule(...args),
```
Comment (khalillabban) [inline __tests__/Schedule.test.tsx]: Same note as above; using any[] for the mock parameters removes type safety. Typing these mocks using the original function signatures could make the test more resilient to future refactors.

File: app/schedule.tsx [hunk 7]
```diff
@@ -127,6 +129,7 @@ export default function ScheduleScreen() {
   const disconnect = useCallback(async () => {
     try {
       await deleteGoogleAccessToken();
+      await AsyncStorage.removeItem("scheduleItems");
```
Comment (khalillabban) [inline app/schedule.tsx]: Suggestion: since "scheduleItems" is used in multiple places (save/load/remove), extracting it to a shared constant would help prevent inconsistencies if the key ever needs to change.

File: utils/parseCourseEvents.ts [hunk 8]
```diff
@@ -1,44 +1,75 @@
...
+        room,
       };
     })
     .filter((x): x is ScheduleItem => Boolean(x))
     .sort((a, b) => a.start.getTime() - b.start.getTime());
+}
+
+// Save items to AsyncStorage (call this after parseCourseEvents)
+export async function saveSchedule(items: ScheduleItem[]): Promise<void> {
+  await AsyncStorage.setItem("scheduleItems", JSON.stringify(items));
```
Comment (khalillabban) [inline utils/parseCourseEvents.ts]: Suggestion: since "scheduleItems" is also used when loading and clearing the cache, it may be worth extracting it into a shared constant so the storage key only lives in one place.
